domain/secret/service/access.go

```diff
@@ -6,27 +6,119 @@ package service
 import (
 	"context"
 
+	"github.com/juju/errors"
+
 	"github.com/juju/juju/core/secrets"
+	domainsecret "github.com/juju/juju/domain/secret"
 )
 
 func (s *SecretService) GetSecretGrants(ctx context.Context, uri *secrets.URI, role secrets.SecretRole) ([]secrets.AccessInfo, error) {
 	//TODO implement me
 	return nil, nil
 }
 
-func (s *SecretService) GetSecretAccess(ctx context.Context, uri *secrets.URI, consumer SecretAccessor) (secrets.SecretRole, error) {
-	//TODO implement me
-	return secrets.RoleManage, nil
+// GetSecretAccessScope returns the access scope for the specified accessor's permission on the secret.
+// It returns an error satisfying [secreterrors.SecretNotFound] if the secret is not found.
+func (s *SecretService) GetSecretAccessScope(ctx context.Context, uri *secrets.URI, accessor SecretAccessor) (SecretAccessScope, error) {
+	ap := domainsecret.AccessParams{
+		SubjectID: accessor.ID,
+	}
+	switch accessor.Kind {
+	case UnitAccessor:
+		ap.SubjectTypeID = domainsecret.SubjectUnit
+	case ApplicationAccessor:
+		ap.SubjectTypeID = domainsecret.SubjectApplication
+	case RemoteApplicationAccessor:
+		ap.SubjectTypeID = domainsecret.SubjectRemoteApplication
+	case ModelAccessor:
+		ap.SubjectTypeID = domainsecret.SubjectModel
+	}
+	accessScope, err := s.st.GetSecretAccessScope(ctx, uri, ap)
+	if err != nil {
+		return SecretAccessScope{}, errors.Trace(err)
+	}
+	result := SecretAccessScope{
+		ID: accessScope.ScopeID,
+	}
+	switch accessScope.ScopeTypeID {
+	case domainsecret.ScopeUnit:
+		result.Kind = UnitAccessScope
+	case domainsecret.ScopeApplication:
+		result.Kind = ApplicationAccessScope
+	case domainsecret.ScopeModel:
+		result.Kind = ModelAccessScope
+	case domainsecret.ScopeRelation:
+		result.Kind = RelationAccessScope
+	}
+	return result, nil
 }
 
-func (s *SecretService) GetSecretAccessScope(ctx context.Context, uri *secrets.URI, consumer SecretAccessor) (SecretAccessScope, error) {
-	//TODO implement me
-	panic("implement me")
+// GetSecretAccess returns the access to the secret for the specified accessor.
+// It returns an error satisfying [secreterrors.SecretNotFound] if the secret is not found.
+func (s *SecretService) GetSecretAccess(ctx context.Context, uri *secrets.URI, accessor SecretAccessor) (secrets.SecretRole, error) {
+	ap := domainsecret.AccessParams{
+		SubjectID: accessor.ID,
+	}
+	switch accessor.Kind {
+	case UnitAccessor:
+		ap.SubjectTypeID = domainsecret.SubjectUnit
+	case ApplicationAccessor:
+		ap.SubjectTypeID = domainsecret.SubjectApplication
+	case RemoteApplicationAccessor:
+		ap.SubjectTypeID = domainsecret.SubjectRemoteApplication
+	case ModelAccessor:
+		ap.SubjectTypeID = domainsecret.SubjectModel
+	}
+	role, err := s.st.GetSecretAccess(ctx, uri, ap)
+	if err != nil {
+		return secrets.RoleNone, errors.Trace(err)
+	}
+	// "none" is db value, secret enum is "".
+	if role == "none" {
```

Comment:
Do we have a const somewhere?

Reply:
This is the value from the db lookup table. The string is only used in this one place in state where the look up table value is loaded. It's used because the core secrets None role is "".